src/wasm/wasm-type.cpp

```diff
@@ -139,13 +139,34 @@ struct SubTyper {
   bool isSubType(HeapType a, HeapType b);
   bool isSubType(const Tuple& a, const Tuple& b);
   bool isSubType(const Field& a, const Field& b);
-  bool isSubType(const HeapTypeInfo& a, const HeapTypeInfo& b);
   bool isSubType(const Signature& a, const Signature& b);
   bool isSubType(const Struct& a, const Struct& b);
   bool isSubType(const Array& a, const Array& b);
   bool isSubType(const Rtt& a, const Rtt& b);
 };
 
+// Helper for finding the equirecursive least upper bound of two types.
+struct TypeBounder {
+  TypeBuilder builder;
+  // The indices in `builder` at which the LUB of each pair of HeapTypes are
+  // being constructed.
+  std::unordered_map<std::pair<HeapType, HeapType>, size_t> indices;
+
+  bool hasLeastUpperBound(Type a, Type b);
+  Type getLeastUpperBound(Type a, Type b);
+
+  // Return true and set `out` to be the LUB iff a LUB was found. The HeapType
+  // and Struct overloads are exceptional because they are infallible.
```

Comment:
I'd expand this comment to add something like "because `anyref` is a supertype of all HeapTypes etc." as you mentioned